modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/modules.js

```diff
@@ -14,14 +14,76 @@
 
 import React from 'react';
 
+const moduleURLPromises = {};
+
 export function getLiferayJsModule(moduleURL) {
-	return new Promise((resolve, reject) => {
-		Liferay.Loader.require(
-			moduleURL,
-			(jsModule) => resolve(jsModule.default || jsModule),
-			(error) => reject(error)
-		);
-	});
+	if (moduleURL.includes('|')) {
+		if (!moduleURLPromises[moduleURL]) {
+			moduleURLPromises[moduleURL] = new Promise((resolve, reject) => {
+				// GreenAppleFDSCellRenderer|/o/ce-sample/index.js
+				const parts = moduleURL.split('|');
+
+				// TODO: We may want to create a npm-scripts webpack plugin to add /*webpackIgnore: true*/ to every dynamic import
+				import(/*webpackIgnore: true*/ parts[1])
+					.then((module) => {
+						let renderer = module[parts[0]];
+
+						resolve((props) => {
+							return React.createElement(
+								class extends React.Component {
+									constructor(props) {
+										super(props);
+
+										this.props = props;
+										this.ref = React.createRef();
+									}
+
+									render() {
+										// Sandbox passed-in args
+										const {value} = this.props;
+
+										setTimeout(
+											() => {
+												renderer({
+													element: this.ref.current,
+													value
+												});
+											},
+											0
+										);
+
+										return <div ref={this.ref}></div>;
```

Comment:
If we are only rendering to the element here, why even use React to create an element and a component instance? Couldn't we just pass `resolve((props) => renderer({props}))`?

Reply:
Yeah, maybe. I'm not an expert in React, so I haven't paid real attention to those details. I expected one of you clever guys deciding that :sweat_smile: